File: AngularJS/src/org/angular2/entities/Angular2Declaration.java
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.entities;

import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.Collections;

import static org.angular2.entities.Angular2EntitiesProvider.getDeclarationToModuleMap;

/**
 * Also known as <a href="https://angular.io/guide/glossary#declarable">Declarable</a>.
 * <p>
 *  Common interface of entities that can be added to {@code @NgModule.declarations} array (only if not standalone).
 */
public interface Angular2Declaration extends Angular2Entity {

  default @NotNull Collection<Angular2Module> getAllDeclaringModules() {
    var map = getDeclarationToModuleMap(getSourceElement().getProject());
    return Collections.unmodifiableCollection(map.get(this));
  }
}

File: AngularJS/src/org/angular2/entities/Angular2Declaration.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.entities

import com.intellij.model.Pointer
import org.angular2.entities.Angular2EntitiesProvider.getDeclarationToModuleMap
import java.util.*

/**
 * Also known as [Declarable](https://angular.io/guide/glossary#declarable).
 *
 *
 * Common interface of entities that can be added to `@NgModule.declarations` array (only if not standalone).
 */
interface Angular2Declaration : Angular2Entity {

  val isStandalone: Boolean

  val allDeclaringModules: Collection<Angular2Module>
    get() =
      if (isStandalone)
        emptyList()
      else
        Collections.unmodifiableCollection(getDeclarationToModuleMap(sourceElement.project).get(this))

  override fun createPointer(): Pointer<out Angular2Declaration>
}
